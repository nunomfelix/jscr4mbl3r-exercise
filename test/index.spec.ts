import { spawn } from 'child_process';
import fs from 'fs';
import path from 'path';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';

const mock = new MockAdapter(axios);

const fakeHTML = `
<html>
<head>
    <title>Test Page</title>
</head>
<body>
    <h1>Hello, world!</h1>
</body>
</html>
`;

describe('CLI', () => {
  beforeAll(() => {
    mock.onGet('https://example.com').reply(200, fakeHTML);
  });

  afterAll(() => {
    mock.reset();
  });

  it('scrapes a URL and writes the output to a file', (done) => {
    // Generate a unique filename for the output file
    const outputFilePath = path.join(__dirname, `output_${Date.now()}.json`);

    // Run the CLI command
    const child = spawn('node', ['dist/index.js', 'https://example.com', '-o', outputFilePath]);

    child.on('error', (err) => {
      // If there's an error, fail the test
      done(err);
    });

    child.on('exit', (code) => {
      if (code !== 0) {
        done(new Error(`Process exited with code ${code}`));
      } else {
        // Check that the output file was created
        const fileExists = fs.existsSync(outputFilePath);
        expect(fileExists).toBe(true);

        if (fileExists) {
          // Check that the file content is as expected
          const fileContent = fs.readFileSync(outputFilePath, 'utf8');
          expect(JSON.parse(fileContent)).toMatchSnapshot();
          
          // Delete the file after test
          fs.unlinkSync(outputFilePath);
        }

        done();
      }
    });
  });
});
