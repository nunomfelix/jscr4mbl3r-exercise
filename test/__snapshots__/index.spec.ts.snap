// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI scrapes a URL and writes the output to a file 1`] = `
{
  "maxDepth": 4,
  "resources": {
    "https://www.iana.org/domains/example": 1,
  },
  "tags": {
    "a": {
      "attributes": {
        "href": 1,
      },
      "children": {},
      "count": 1,
    },
    "body": {
      "attributes": {},
      "children": {
        "div": 1,
      },
      "count": 1,
    },
    "div": {
      "attributes": {},
      "children": {
        "h1": 1,
        "p": 2,
      },
      "count": 1,
    },
    "h1": {
      "attributes": {},
      "children": {},
      "count": 1,
    },
    "head": {
      "attributes": {},
      "children": {
        "meta": 3,
        "title": 1,
      },
      "count": 1,
    },
    "html": {
      "attributes": {},
      "children": {
        "body": 1,
        "head": 1,
      },
      "count": 1,
    },
    "meta": {
      "attributes": {
        "charset": 1,
        "content": 2,
        "http-equiv": 1,
        "name": 1,
      },
      "children": {},
      "count": 3,
    },
    "p": {
      "attributes": {},
      "children": {
        "a": 1,
      },
      "count": 2,
    },
    "title": {
      "attributes": {},
      "children": {},
      "count": 1,
    },
  },
}
`;
